### DISCUSSION TITLE
Implicits and Overloading

### DISCUSSION THREAD

**Discussion by TimWhiting** [koka-lang/koka#431] [General]
This is a general discussion topic to discuss the new overloading and implicits feature in Koka. Daan and I worked hard to make it feel really nice to the user, while adhering to Koka's [min-gen](https://koka-lang.github.io/koka/doc/book.html#why-mingen) design (Daan really did all the work, I just gave some feedback). See the new samples for more information on the design of the feature: [qualifiers / overloading](https://github.com/koka-lang/koka/blob/dev/samples/syntax/qualifiers.kk), [implicits](https://github.com/koka-lang/koka/blob/dev/samples/syntax/implicits.kk). Happy to take any feedback, so we can all contribute to making Koka the best min-gen functional language.

**Comment by woojamon** [comment]
I find these features to be done very well so far. But what really makes these features for me are the hint inlays in vscode.

**Qualifiers/Overloading:**

I find the implementation here scratches an itch I've felt in C# for years, where in the case of extension method ambiguity there's no way to qualify/disambiguate the desired method inline. I love that in Koka I can qualify with the module name in a "path-like" way *inline*. And if the compiler is able to resolve the ambiguity for me, I can just CTRL+OPTION on my mac to see the inlay hint of where it found the resolution.

Coming the from C# concept of overloading, though, it doesn't make sense to me why the following doesn't compile, since there wouldn't be any ambiguity at the call site.
```koka
fun foo()
    "bar"

fun foo(bar:string)
    bar
```
Of course, C# also doesn't have first-class functions, so `val f = foo` would be ambiguous to the compiler, at least until `f` is applied with either `()` or some string. But some languages (like F#) seem to fine with some ambiguities as long as the compiler can resolve the issue *somewhere* in the code base. (For example, if I had `line 1: val f = foo` and `line 5: val x = f()`, then the compiler would know back on line 1 which overload to resolve. But then if I had `line 6: var y = f("bar")` there would be a compiler error due to ambiguity again.)

**Implicits:**
Generally I'm not a fan of hidden "magic" parameters (thanks to codebases with much-abused C#), but again here, at least I can CTRL+OPTION to discover what magic is being applied to my code. And I do see how it can reduce boilerplate parameters and make the code more readable, so overall I think this the implicits feature is done well.

**Reply by TimWhiting** [reply]
```koka
fun foo()
    "bar"

fun foo(bar:string)
    bar
```
I think you precisely pointed out the problem here: `val f = foo` would be ambiguous to the compiler, at least until `f` is applied with either `()` or some string. The problem with this is that to do this properly (theoretically sound and cover every case) you have to do a full control flow analysis (which is O(n^3) runtime at minimum). There are shortcuts, such as requiring first class function parameters to always be annotated, or not supporting tracing the flow for first class function parameters, but they quickly become complex rules, and make the compiler unnecessarily complex - `F#` likely has some complex rules or unspecified compiler hacks to make this work in a majority of cases. 

With just a few people maintaining the compiler we aim to give you the best user experience while keeping both the language and compiler simple. I recommend just changing your example to:

```koka
fun unit/foo()
    "bar"

fun string/foo(bar:string)
    bar
```

This allows you to still just use `foo()` and let the compiler resolve the overloading for the majority of your use cases (when applying it directly), but makes it so that you can also refer to the identifier as a first class value unambiguously (i.e. when using it as `val f = unit/foo`). This has the side advantage that when you do use it as a first class value it is also clear to the user which `foo` you are referring to. This means the compiler has to do fewer 'magic' tricks or hacks, and makes overloading very principled.

Thanks for all of the positive feedback! We really worked hard to make the language server really well integrated with these features, and are glad to see the work paid off.
